# Importing-Preprocessing Dataset

In [1]:
import pandas as pd
from sklearn.impute import KNNImputer
from sklearn.model_selection import train_test_split

samples = pd.read_csv('wines.csv')
samples['type'] = samples['type'].replace({'white': 0, 'red': 1}).astype(int)
samples[:] = KNNImputer(n_neighbors=25, weights='distance').fit_transform(samples)

samples['other sulfur dioxide'] = samples['total sulfur dioxide'] - samples['free sulfur dioxide']

targets = samples['quality']
samples.drop(columns=['quality', 'total sulfur dioxide'], inplace=True)

x_train, x_test, y_train, y_test = train_test_split(samples, targets, test_size=0.1, random_state=0)
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((5847, 12), (5847,), (650, 12), (650,))

# Define Scaling Pipelines

In [2]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler, RobustScaler, Normalizer

scaling_methods = {
    'None': None,
    'MinMax': MinMaxScaler(),
    'StandardScaler': StandardScaler(),
    'RobustScaler': RobustScaler(),
    'Normalizer': Normalizer()
}

# Training a Neural Network

In [3]:
import tensorflow as tf
import tensorflow_addons as tfa

tf.random.set_seed(0)


def build_model(input_size):
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Input(input_size))
    model.add(tf.keras.layers.Dense(units=256, activation='gelu', use_bias=False))
    model.add(tf.keras.layers.BatchNormalization())
    model.add(tf.keras.layers.Dense(units=128, activation='gelu'))
    model.add(tf.keras.layers.Dense(1))

    model.compile(optimizer=tfa.optimizers.Yogi(learning_rate=0.0005), loss='mae', metrics=['mae', 'mape'])

    return model


model = build_model(samples.shape[1:])
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 256)               3072      
                                                                 
 batch_normalization (BatchN  (None, 256)              1024      
 ormalization)                                                   
                                                                 
 dense_1 (Dense)             (None, 128)               32896     
                                                                 
 dense_2 (Dense)             (None, 1)                 129       
                                                                 
Total params: 37,121
Trainable params: 36,609
Non-trainable params: 512
_________________________________________________________________


In [4]:
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error

epochs = 300
batch_size = 32
early_stopping = tf.keras.callbacks.EarlyStopping(
    monitor='val_loss',
    mode='min',
    patience=30
)

for scaling_method, scaler in scaling_methods.items():
    if scaler is not None:
        scaled_x_train = scaler.fit_transform(x_train)
        scaled_x_test = scaler.transform(x_test)
    else:
        scaled_x_train = x_train
        scaled_x_test = x_test

    model = build_model(x_train.shape[1:])
    model.fit(
        scaled_x_train,
        y_train,
        batch_size=batch_size,
        epochs=epochs,
        callbacks=[early_stopping],
        validation_data=(scaled_x_test, y_test),
        verbose=0
    )
    y_pred = model.predict(scaled_x_test)

    mae = mean_absolute_error(y_test, y_pred)
    mape = mean_absolute_percentage_error(y_test, y_pred)

    print(f'MAE: {mae}\tMAPE: {mape}\tScaling Method: {scaling_method}')

21/21 [==============================] - 0s 848us/step
MAE: 0.574664941567641	MAPE: 0.10425148522548187	Scaling Method: None
21/21 [==============================] - 0s 798us/step
MAE: 0.5335880668346699	MAPE: 0.09286738122164548	Scaling Method: MinMax
21/21 [==============================] - 0s 798us/step
MAE: 0.5073002829918495	MAPE: 0.08849438374645108	Scaling Method: StandardScaler
21/21 [==============================] - 0s 798us/step
MAE: 0.5373223462471595	MAPE: 0.09617116946877141	Scaling Method: RobustScaler
21/21 [==============================] - 0s 848us/step
MAE: 0.6214593729606042	MAPE: 0.11294098635645576	Scaling Method: Normalizer


| Scaling        | Mean Absolute Error **(MAE)** | Mean Absolute Percentage Error **(MAPE)** |
|----------------|-------------------------------|-------------------------------------------|
| No Scaling     | 0.5747                        | 0.1043                                    |
| MinMax         | 0.5336                        | 0.0923                                    |
| StandardScaler | **0.5073**                    | **0.0885**                                |
| RobustScaler   | 0.5373                        | 0.0962                                    |
| Normalizer     | 0.6215                        | 0.1129                                    |

# Improving the Regression Model

Obviously, the red wine has different properties than the white wine. We can use this to scale differently
the 2 types of wine.

In [5]:
white_wine_ind = samples['type'] == 0.0
red_wine_ind = samples['type'] == 1.0

for scaling_method in scaling_methods:
    if scaling_method != 'None':
        scaled_x_train = x_train.copy()
        scaled_x_test = x_test.copy()

        white_scaler = scaling_methods[scaling_method]
        scaled_x_train.loc[white_wine_ind] = white_scaler.fit_transform(scaled_x_train.loc[white_wine_ind])
        scaled_x_test.loc[white_wine_ind] = white_scaler.transform(scaled_x_test.loc[white_wine_ind])

        red_scaler = scaling_methods[scaling_method]
        scaled_x_train.loc[red_wine_ind] = red_scaler.fit_transform(scaled_x_train.loc[red_wine_ind])
        scaled_x_test.loc[red_wine_ind] = red_scaler.transform(scaled_x_test.loc[red_wine_ind])

        model = build_model(x_train.shape[1:])
        model.fit(
            scaled_x_train,
            y_train,
            batch_size=batch_size,
            epochs=epochs,
            callbacks=[early_stopping],
            validation_data=(scaled_x_test, y_test),
            verbose=0
        )
        y_pred = model.predict(scaled_x_test)

        mae = mean_absolute_error(y_test, y_pred)
        mape = mean_absolute_percentage_error(y_test, y_pred)

        print(f'MAE: {mae}\tMAPE: {mape}\tScaling Method: {scaling_method}')

21/21 [==============================] - 0s 848us/step
MAE: 0.5506982950063852	MAPE: 0.09579753529021155	Scaling Method: MinMax
21/21 [==============================] - 0s 758us/step
MAE: 0.5123593665544803	MAPE: 0.09001121112475029	Scaling Method: StandardScaler
21/21 [==============================] - 0s 868us/step
MAE: 0.5049129166969886	MAPE: 0.08913842107612135	Scaling Method: RobustScaler
21/21 [==============================] - 0s 798us/step
MAE: 0.5900471496582032	MAPE: 0.10255494306201027	Scaling Method: Normalizer


| Scaling        | Mean Absolute Error **(MAE)** | Mean Absolute Percentage Error **(MAPE)** |
|----------------|-------------------------------|-------------------------------------------|
| No Scaling     | 0.5747                        | 0.1043                                    |
| MinMax         | 0.5507                        | 0.0958                                    |
| StandardScaler | 0.5124                        | 0.0900                                    |
| RobustScaler   | **0.5049**                    | **0.0891**                                |
| Normalizer     | 0.5900                        | 0.1026                                    |

We may notice that by using separate scalers for white and red wine samples the overal MAE was reduced

# Hyper-Parameter Tuning with Optuna

In [4]:
import optuna

white_wine_ind = samples['type'] == 0.0
red_wine_ind = samples['type'] == 1.0

scaler = StandardScaler()
x_train.loc[white_wine_ind] = scaler.fit_transform(x_train.loc[white_wine_ind])
x_test.loc[white_wine_ind] = scaler.transform(x_test.loc[white_wine_ind])

scaler = StandardScaler()
x_train.loc[red_wine_ind] = scaler.fit_transform(x_train.loc[red_wine_ind])
x_test.loc[red_wine_ind] = scaler.transform(x_test.loc[red_wine_ind])

In [8]:
def create_model(trial):
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Input(x_train.shape[1:]))

    n_layers = trial.suggest_int('n_layers', 1, 3)
    for i in range(n_layers - 1):
        hidden_units = trial.suggest_int(f'hidden_units_{i}', 32, 256)
        use_batch_norm = bool(trial.suggest_int(f'use_batch_norm_{i}', 0, 1))
        use_bias = False if use_batch_norm else True

        model.add(tf.keras.layers.Dense(units=hidden_units, activation='gelu', use_bias=use_bias))
        if use_batch_norm:
            model.add(tf.keras.layers.BatchNormalization())

    hidden_units = trial.suggest_int('hidden_units_last', 32, 256)
    model.add(tf.keras.layers.Dense(units=hidden_units, activation='gelu'))
    model.add(tf.keras.layers.Dense(units=1))

    optimizer = trial.suggest_categorical('optimizer', ['Yogi', 'AdamW'])
    loss = trial.suggest_categorical('loss', ['mse', 'mae', 'mape'])

    if optimizer == 'Adam':
        optim = tf.keras.optimizers.Adam()
    elif optimizer == 'Yogi':
        optim = tfa.optimizers.Yogi()
    else:
        optim = tfa.optimizers.AdamW(weight_decay=0.01)

    model.compile(optimizer=optim, loss=loss, metrics=[loss])

    return model


def objective(trial):
    model = create_model(trial)

    if trial.should_prune():
        raise optuna.TrialPruned()

    model.fit(
        x_train,
        y_train,
        batch_size=32,
        epochs=100,
        callbacks=[early_stopping],
        validation_data=(x_test, y_test),
        verbose=0
    )

    y_pred = model.predict(x_test)
    return mean_absolute_error(y_test, y_pred)


n_trials = 100
study = optuna.create_study(direction='minimize')
study.optimize(objective, n_trials=n_trials)
study.trials_dataframe()

[I 2022-07-15 18:07:44,977] A new study created in memory with name: no-name-20b38cfc-5d38-4c7e-9700-94c70a8a35ef


21/21 [==============================] - 0s 748us/step


[I 2022-07-15 18:08:40,118] Trial 0 finished with value: 0.5385948093120868 and parameters: {'n_layers': 2, 'hidden_units_0': 169, 'use_batch_norm_0': 0, 'hidden_units_last': 208, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 0 with value: 0.5385948093120868.


21/21 [==============================] - 0s 720us/step


[I 2022-07-15 18:09:06,697] Trial 1 finished with value: 5.213332895584978 and parameters: {'n_layers': 1, 'hidden_units_last': 44, 'optimizer': 'AdamW', 'loss': 'mape'}. Best is trial 0 with value: 0.5385948093120868.


21/21 [==============================] - 0s 698us/step


[I 2022-07-15 18:09:19,101] Trial 2 finished with value: 1.8600141461078936 and parameters: {'n_layers': 2, 'hidden_units_0': 209, 'use_batch_norm_0': 0, 'hidden_units_last': 51, 'optimizer': 'AdamW', 'loss': 'mape'}. Best is trial 0 with value: 0.5385948093120868.


21/21 [==============================] - 0s 648us/step


[I 2022-07-15 18:10:02,722] Trial 3 finished with value: 0.531550312775832 and parameters: {'n_layers': 1, 'hidden_units_last': 156, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 3 with value: 0.531550312775832.


21/21 [==============================] - 0s 898us/step


[I 2022-07-15 18:10:24,834] Trial 4 finished with value: 0.662501441515409 and parameters: {'n_layers': 3, 'hidden_units_0': 74, 'use_batch_norm_0': 1, 'hidden_units_1': 145, 'use_batch_norm_1': 0, 'hidden_units_last': 126, 'optimizer': 'AdamW', 'loss': 'mae'}. Best is trial 3 with value: 0.531550312775832.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 18:10:51,178] Trial 5 finished with value: 0.6493547835716834 and parameters: {'n_layers': 3, 'hidden_units_0': 95, 'use_batch_norm_0': 0, 'hidden_units_1': 254, 'use_batch_norm_1': 0, 'hidden_units_last': 90, 'optimizer': 'AdamW', 'loss': 'mse'}. Best is trial 3 with value: 0.531550312775832.


21/21 [==============================] - 0s 764us/step


[I 2022-07-15 18:11:11,346] Trial 6 finished with value: 0.653750985952524 and parameters: {'n_layers': 2, 'hidden_units_0': 167, 'use_batch_norm_0': 0, 'hidden_units_last': 223, 'optimizer': 'AdamW', 'loss': 'mape'}. Best is trial 3 with value: 0.531550312775832.


21/21 [==============================] - 0s 708us/step


[I 2022-07-15 18:11:49,558] Trial 7 finished with value: 0.5460694888921884 and parameters: {'n_layers': 1, 'hidden_units_last': 60, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 3 with value: 0.531550312775832.


21/21 [==============================] - 0s 711us/step


[I 2022-07-15 18:12:16,296] Trial 8 finished with value: 0.5258808257029607 and parameters: {'n_layers': 2, 'hidden_units_0': 223, 'use_batch_norm_0': 0, 'hidden_units_last': 70, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 8 with value: 0.5258808257029607.


21/21 [==============================] - 0s 648us/step


[I 2022-07-15 18:12:51,474] Trial 9 finished with value: 0.5595147195229163 and parameters: {'n_layers': 1, 'hidden_units_last': 43, 'optimizer': 'Yogi', 'loss': 'mape'}. Best is trial 8 with value: 0.5258808257029607.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 18:14:00,614] Trial 10 finished with value: 0.5148054966559776 and parameters: {'n_layers': 3, 'hidden_units_0': 243, 'use_batch_norm_0': 1, 'hidden_units_1': 43, 'use_batch_norm_1': 1, 'hidden_units_last': 114, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 10 with value: 0.5148054966559776.


21/21 [==============================] - 0s 898us/step


[I 2022-07-15 18:15:24,855] Trial 11 finished with value: 0.509449208332942 and parameters: {'n_layers': 3, 'hidden_units_0': 254, 'use_batch_norm_0': 1, 'hidden_units_1': 39, 'use_batch_norm_1': 1, 'hidden_units_last': 115, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 11 with value: 0.509449208332942.


21/21 [==============================] - 0s 947us/step


[I 2022-07-15 18:16:47,579] Trial 12 finished with value: 0.5033799384190486 and parameters: {'n_layers': 3, 'hidden_units_0': 256, 'use_batch_norm_0': 1, 'hidden_units_1': 32, 'use_batch_norm_1': 1, 'hidden_units_last': 125, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 12 with value: 0.5033799384190486.


21/21 [==============================] - 0s 898us/step


[I 2022-07-15 18:18:07,547] Trial 13 finished with value: 0.5230167000110333 and parameters: {'n_layers': 3, 'hidden_units_0': 242, 'use_batch_norm_0': 1, 'hidden_units_1': 32, 'use_batch_norm_1': 1, 'hidden_units_last': 171, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 12 with value: 0.5033799384190486.


21/21 [==============================] - 0s 898us/step


[I 2022-07-15 18:19:28,704] Trial 14 finished with value: 0.4855645022025475 and parameters: {'n_layers': 3, 'hidden_units_0': 196, 'use_batch_norm_0': 1, 'hidden_units_1': 105, 'use_batch_norm_1': 1, 'hidden_units_last': 105, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 14 with value: 0.4855645022025475.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 18:20:48,724] Trial 15 finished with value: 0.49853695942805365 and parameters: {'n_layers': 3, 'hidden_units_0': 191, 'use_batch_norm_0': 1, 'hidden_units_1': 118, 'use_batch_norm_1': 1, 'hidden_units_last': 176, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 14 with value: 0.4855645022025475.


21/21 [==============================] - 0s 874us/step


[I 2022-07-15 18:21:33,417] Trial 16 finished with value: 0.5443208987896259 and parameters: {'n_layers': 3, 'hidden_units_0': 121, 'use_batch_norm_0': 1, 'hidden_units_1': 117, 'use_batch_norm_1': 1, 'hidden_units_last': 192, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 14 with value: 0.4855645022025475.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 18:22:14,020] Trial 17 finished with value: 0.5320751358912541 and parameters: {'n_layers': 2, 'hidden_units_0': 188, 'use_batch_norm_0': 1, 'hidden_units_last': 236, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 14 with value: 0.4855645022025475.


21/21 [==============================] - 0s 947us/step


[I 2022-07-15 18:23:35,069] Trial 18 finished with value: 0.5081002063017626 and parameters: {'n_layers': 3, 'hidden_units_0': 35, 'use_batch_norm_0': 1, 'hidden_units_1': 122, 'use_batch_norm_1': 1, 'hidden_units_last': 187, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 14 with value: 0.4855645022025475.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 18:24:39,882] Trial 19 finished with value: 0.47947508041675274 and parameters: {'n_layers': 3, 'hidden_units_0': 143, 'use_batch_norm_0': 1, 'hidden_units_1': 163, 'use_batch_norm_1': 0, 'hidden_units_last': 152, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 19 with value: 0.47947508041675274.


21/21 [==============================] - 0s 803us/step


[I 2022-07-15 18:25:15,952] Trial 20 finished with value: 0.5414736777085525 and parameters: {'n_layers': 2, 'hidden_units_0': 130, 'use_batch_norm_0': 1, 'hidden_units_last': 89, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 19 with value: 0.47947508041675274.


21/21 [==============================] - 0s 841us/step


[I 2022-07-15 18:26:27,009] Trial 21 finished with value: 0.46377170159266545 and parameters: {'n_layers': 3, 'hidden_units_0': 160, 'use_batch_norm_0': 1, 'hidden_units_1': 174, 'use_batch_norm_1': 0, 'hidden_units_last': 156, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 918us/step


[I 2022-07-15 18:27:38,202] Trial 22 finished with value: 0.5038847086979793 and parameters: {'n_layers': 3, 'hidden_units_0': 153, 'use_batch_norm_0': 1, 'hidden_units_1': 182, 'use_batch_norm_1': 0, 'hidden_units_last': 152, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 879us/step


[I 2022-07-15 18:28:47,930] Trial 23 finished with value: 0.4739088564652663 and parameters: {'n_layers': 3, 'hidden_units_0': 142, 'use_batch_norm_0': 1, 'hidden_units_1': 183, 'use_batch_norm_1': 0, 'hidden_units_last': 140, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 18:29:38,499] Trial 24 finished with value: 0.5155580065800593 and parameters: {'n_layers': 3, 'hidden_units_0': 108, 'use_batch_norm_0': 1, 'hidden_units_1': 191, 'use_batch_norm_1': 0, 'hidden_units_last': 136, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 18:30:25,782] Trial 25 finished with value: 0.5246667942633996 and parameters: {'n_layers': 2, 'hidden_units_0': 146, 'use_batch_norm_0': 1, 'hidden_units_last': 158, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 18:30:47,417] Trial 26 finished with value: 0.5986809459099403 and parameters: {'n_layers': 3, 'hidden_units_0': 161, 'use_batch_norm_0': 1, 'hidden_units_1': 188, 'use_batch_norm_1': 0, 'hidden_units_last': 146, 'optimizer': 'AdamW', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 849us/step


[I 2022-07-15 18:31:44,018] Trial 27 finished with value: 0.4786649960737962 and parameters: {'n_layers': 3, 'hidden_units_0': 132, 'use_batch_norm_0': 1, 'hidden_units_1': 229, 'use_batch_norm_1': 0, 'hidden_units_last': 169, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 794us/step


[I 2022-07-15 18:32:42,834] Trial 28 finished with value: 0.49806565358088567 and parameters: {'n_layers': 2, 'hidden_units_0': 86, 'use_batch_norm_0': 1, 'hidden_units_last': 200, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 866us/step


[I 2022-07-15 18:33:53,767] Trial 29 finished with value: 0.48082441843472995 and parameters: {'n_layers': 3, 'hidden_units_0': 127, 'use_batch_norm_0': 1, 'hidden_units_1': 230, 'use_batch_norm_1': 0, 'hidden_units_last': 220, 'optimizer': 'Yogi', 'loss': 'mape'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 709us/step


[I 2022-07-15 18:34:36,087] Trial 30 finished with value: 0.5161285807536199 and parameters: {'n_layers': 2, 'hidden_units_0': 60, 'use_batch_norm_0': 0, 'hidden_units_last': 177, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 18:35:45,978] Trial 31 finished with value: 0.4797979208139273 and parameters: {'n_layers': 3, 'hidden_units_0': 142, 'use_batch_norm_0': 1, 'hidden_units_1': 162, 'use_batch_norm_1': 0, 'hidden_units_last': 165, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 898us/step


[I 2022-07-15 18:36:57,947] Trial 32 finished with value: 0.46675801277160645 and parameters: {'n_layers': 3, 'hidden_units_0': 113, 'use_batch_norm_0': 1, 'hidden_units_1': 212, 'use_batch_norm_1': 0, 'hidden_units_last': 139, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 923us/step


[I 2022-07-15 18:38:10,139] Trial 33 finished with value: 0.4826749262442956 and parameters: {'n_layers': 3, 'hidden_units_0': 114, 'use_batch_norm_0': 1, 'hidden_units_1': 218, 'use_batch_norm_1': 0, 'hidden_units_last': 133, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 18:38:31,330] Trial 34 finished with value: 0.7178988709816566 and parameters: {'n_layers': 3, 'hidden_units_0': 180, 'use_batch_norm_0': 1, 'hidden_units_1': 213, 'use_batch_norm_1': 0, 'hidden_units_last': 138, 'optimizer': 'AdamW', 'loss': 'mape'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 895us/step


[I 2022-07-15 18:39:41,866] Trial 35 finished with value: 0.4875486318881695 and parameters: {'n_layers': 3, 'hidden_units_0': 173, 'use_batch_norm_0': 1, 'hidden_units_1': 243, 'use_batch_norm_1': 0, 'hidden_units_last': 256, 'optimizer': 'Yogi', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 897us/step


[I 2022-07-15 18:40:00,484] Trial 36 finished with value: 0.6333594564291147 and parameters: {'n_layers': 3, 'hidden_units_0': 100, 'use_batch_norm_0': 1, 'hidden_units_1': 207, 'use_batch_norm_1': 0, 'hidden_units_last': 101, 'optimizer': 'AdamW', 'loss': 'mse'}. Best is trial 21 with value: 0.46377170159266545.


21/21 [==============================] - 0s 797us/step


[I 2022-07-15 18:40:59,867] Trial 37 finished with value: 0.45770603840167706 and parameters: {'n_layers': 3, 'hidden_units_0': 131, 'use_batch_norm_0': 0, 'hidden_units_1': 199, 'use_batch_norm_1': 0, 'hidden_units_last': 186, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 37 with value: 0.45770603840167706.


21/21 [==============================] - 0s 1ms/step


[I 2022-07-15 18:41:13,434] Trial 38 finished with value: 0.7636546171628512 and parameters: {'n_layers': 2, 'hidden_units_0': 82, 'use_batch_norm_0': 0, 'hidden_units_last': 210, 'optimizer': 'AdamW', 'loss': 'mae'}. Best is trial 37 with value: 0.45770603840167706.


21/21 [==============================] - 0s 670us/step


[I 2022-07-15 18:41:52,955] Trial 39 finished with value: 0.5373619101597713 and parameters: {'n_layers': 1, 'hidden_units_last': 125, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 37 with value: 0.45770603840167706.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 18:42:52,815] Trial 40 finished with value: 0.4768749501154973 and parameters: {'n_layers': 3, 'hidden_units_0': 157, 'use_batch_norm_0': 0, 'hidden_units_1': 163, 'use_batch_norm_1': 0, 'hidden_units_last': 188, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 37 with value: 0.45770603840167706.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 18:43:54,124] Trial 41 finished with value: 0.4326023167830247 and parameters: {'n_layers': 3, 'hidden_units_0': 160, 'use_batch_norm_0': 0, 'hidden_units_1': 169, 'use_batch_norm_1': 0, 'hidden_units_last': 191, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 41 with value: 0.4326023167830247.


21/21 [==============================] - 0s 821us/step


[I 2022-07-15 18:44:54,243] Trial 42 finished with value: 0.4614516261907724 and parameters: {'n_layers': 3, 'hidden_units_0': 138, 'use_batch_norm_0': 0, 'hidden_units_1': 198, 'use_batch_norm_1': 0, 'hidden_units_last': 203, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 41 with value: 0.4326023167830247.


21/21 [==============================] - 0s 810us/step


[I 2022-07-15 18:45:53,957] Trial 43 finished with value: 0.449553620998676 and parameters: {'n_layers': 3, 'hidden_units_0': 207, 'use_batch_norm_0': 0, 'hidden_units_1': 204, 'use_batch_norm_1': 0, 'hidden_units_last': 207, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 41 with value: 0.4326023167830247.


21/21 [==============================] - 0s 801us/step


[I 2022-07-15 18:46:53,554] Trial 44 finished with value: 0.4518682986039382 and parameters: {'n_layers': 3, 'hidden_units_0': 213, 'use_batch_norm_0': 0, 'hidden_units_1': 197, 'use_batch_norm_1': 0, 'hidden_units_last': 205, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 41 with value: 0.4326023167830247.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 18:47:56,828] Trial 45 finished with value: 0.4388093354151799 and parameters: {'n_layers': 3, 'hidden_units_0': 223, 'use_batch_norm_0': 0, 'hidden_units_1': 199, 'use_batch_norm_1': 0, 'hidden_units_last': 233, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 41 with value: 0.4326023167830247.


21/21 [==============================] - 0s 807us/step


[I 2022-07-15 18:48:12,503] Trial 46 finished with value: 0.6026899003982544 and parameters: {'n_layers': 3, 'hidden_units_0': 217, 'use_batch_norm_0': 0, 'hidden_units_1': 144, 'use_batch_norm_1': 0, 'hidden_units_last': 239, 'optimizer': 'AdamW', 'loss': 'mae'}. Best is trial 41 with value: 0.4326023167830247.


21/21 [==============================] - 0s 772us/step


[I 2022-07-15 18:49:17,147] Trial 47 finished with value: 0.45111532284663275 and parameters: {'n_layers': 3, 'hidden_units_0': 229, 'use_batch_norm_0': 0, 'hidden_units_1': 199, 'use_batch_norm_1': 0, 'hidden_units_last': 217, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 41 with value: 0.4326023167830247.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 18:50:19,975] Trial 48 finished with value: 0.4324150411899273 and parameters: {'n_layers': 3, 'hidden_units_0': 229, 'use_batch_norm_0': 0, 'hidden_units_1': 230, 'use_batch_norm_1': 0, 'hidden_units_last': 218, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 48 with value: 0.4324150411899273.


21/21 [==============================] - 0s 909us/step


[I 2022-07-15 18:51:24,172] Trial 49 finished with value: 0.465242174588717 and parameters: {'n_layers': 3, 'hidden_units_0': 230, 'use_batch_norm_0': 0, 'hidden_units_1': 255, 'use_batch_norm_1': 0, 'hidden_units_last': 222, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 48 with value: 0.4324150411899273.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 18:52:05,422] Trial 50 finished with value: 0.535535007256728 and parameters: {'n_layers': 1, 'hidden_units_last': 235, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 48 with value: 0.4324150411899273.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 18:53:01,396] Trial 51 finished with value: 0.480418273302225 and parameters: {'n_layers': 3, 'hidden_units_0': 212, 'use_batch_norm_0': 0, 'hidden_units_1': 234, 'use_batch_norm_1': 0, 'hidden_units_last': 211, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 48 with value: 0.4324150411899273.


21/21 [==============================] - 0s 799us/step


[I 2022-07-15 18:54:04,753] Trial 52 finished with value: 0.4422108477812547 and parameters: {'n_layers': 3, 'hidden_units_0': 203, 'use_batch_norm_0': 0, 'hidden_units_1': 219, 'use_batch_norm_1': 0, 'hidden_units_last': 250, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 48 with value: 0.4324150411899273.


21/21 [==============================] - 0s 855us/step


[I 2022-07-15 18:55:08,132] Trial 53 finished with value: 0.44056106127225436 and parameters: {'n_layers': 3, 'hidden_units_0': 234, 'use_batch_norm_0': 0, 'hidden_units_1': 222, 'use_batch_norm_1': 0, 'hidden_units_last': 249, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 48 with value: 0.4324150411899273.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 18:56:12,085] Trial 54 finished with value: 0.44575471804692196 and parameters: {'n_layers': 3, 'hidden_units_0': 243, 'use_batch_norm_0': 0, 'hidden_units_1': 225, 'use_batch_norm_1': 0, 'hidden_units_last': 254, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 48 with value: 0.4324150411899273.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 18:57:16,574] Trial 55 finished with value: 0.4111724328994751 and parameters: {'n_layers': 3, 'hidden_units_0': 240, 'use_batch_norm_0': 0, 'hidden_units_1': 219, 'use_batch_norm_1': 0, 'hidden_units_last': 251, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 1ms/step


[I 2022-07-15 18:58:19,261] Trial 56 finished with value: 0.4522848804180439 and parameters: {'n_layers': 3, 'hidden_units_0': 199, 'use_batch_norm_0': 0, 'hidden_units_1': 244, 'use_batch_norm_1': 0, 'hidden_units_last': 245, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 813us/step


[I 2022-07-15 18:59:22,532] Trial 57 finished with value: 0.43234496666834904 and parameters: {'n_layers': 3, 'hidden_units_0': 238, 'use_batch_norm_0': 0, 'hidden_units_1': 245, 'use_batch_norm_1': 0, 'hidden_units_last': 231, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 19:00:15,267] Trial 58 finished with value: 0.4942438371364887 and parameters: {'n_layers': 2, 'hidden_units_0': 251, 'use_batch_norm_0': 0, 'hidden_units_last': 229, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 796us/step


[I 2022-07-15 19:00:30,137] Trial 59 finished with value: 0.613933742229755 and parameters: {'n_layers': 3, 'hidden_units_0': 234, 'use_batch_norm_0': 0, 'hidden_units_1': 242, 'use_batch_norm_1': 0, 'hidden_units_last': 228, 'optimizer': 'AdamW', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 19:01:33,255] Trial 60 finished with value: 0.4639179174716656 and parameters: {'n_layers': 3, 'hidden_units_0': 238, 'use_batch_norm_0': 0, 'hidden_units_1': 237, 'use_batch_norm_1': 0, 'hidden_units_last': 244, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 901us/step


[I 2022-07-15 19:02:38,762] Trial 61 finished with value: 0.4558307537665734 and parameters: {'n_layers': 3, 'hidden_units_0': 222, 'use_batch_norm_0': 0, 'hidden_units_1': 221, 'use_batch_norm_1': 0, 'hidden_units_last': 247, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 900us/step


[I 2022-07-15 19:03:42,364] Trial 62 finished with value: 0.4544347289892343 and parameters: {'n_layers': 3, 'hidden_units_0': 250, 'use_batch_norm_0': 0, 'hidden_units_1': 248, 'use_batch_norm_1': 0, 'hidden_units_last': 231, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 19:04:45,967] Trial 63 finished with value: 0.45301569278423603 and parameters: {'n_layers': 3, 'hidden_units_0': 203, 'use_batch_norm_0': 0, 'hidden_units_1': 221, 'use_batch_norm_1': 0, 'hidden_units_last': 249, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 19:05:51,222] Trial 64 finished with value: 0.4546359062194824 and parameters: {'n_layers': 3, 'hidden_units_0': 224, 'use_batch_norm_0': 0, 'hidden_units_1': 175, 'use_batch_norm_1': 0, 'hidden_units_last': 240, 'optimizer': 'Yogi', 'loss': 'mape'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 19:06:54,747] Trial 65 finished with value: 0.46254398492666393 and parameters: {'n_layers': 3, 'hidden_units_0': 245, 'use_batch_norm_0': 0, 'hidden_units_1': 90, 'use_batch_norm_1': 0, 'hidden_units_last': 256, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 835us/step


[I 2022-07-15 19:07:59,367] Trial 66 finished with value: 0.4484183641580435 and parameters: {'n_layers': 3, 'hidden_units_0': 220, 'use_batch_norm_0': 0, 'hidden_units_1': 255, 'use_batch_norm_1': 0, 'hidden_units_last': 216, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 898us/step


[I 2022-07-15 19:09:13,340] Trial 67 finished with value: 0.49014960527420043 and parameters: {'n_layers': 3, 'hidden_units_0': 187, 'use_batch_norm_0': 0, 'hidden_units_1': 212, 'use_batch_norm_1': 1, 'hidden_units_last': 225, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 847us/step


[I 2022-07-15 19:10:14,412] Trial 68 finished with value: 0.4517338576683631 and parameters: {'n_layers': 3, 'hidden_units_0': 236, 'use_batch_norm_0': 0, 'hidden_units_1': 227, 'use_batch_norm_1': 0, 'hidden_units_last': 236, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 19:11:15,492] Trial 69 finished with value: 0.43709468071277324 and parameters: {'n_layers': 3, 'hidden_units_0': 256, 'use_batch_norm_0': 0, 'hidden_units_1': 232, 'use_batch_norm_1': 0, 'hidden_units_last': 196, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 748us/step


[I 2022-07-15 19:11:46,398] Trial 70 finished with value: 0.529197031901433 and parameters: {'n_layers': 2, 'hidden_units_0': 256, 'use_batch_norm_0': 0, 'hidden_units_last': 199, 'optimizer': 'Yogi', 'loss': 'mape'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 19:12:29,574] Trial 71 finished with value: 0.47813629663907564 and parameters: {'n_layers': 3, 'hidden_units_0': 244, 'use_batch_norm_0': 0, 'hidden_units_1': 232, 'use_batch_norm_1': 0, 'hidden_units_last': 250, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 19:13:32,841] Trial 72 finished with value: 0.4648621074969952 and parameters: {'n_layers': 3, 'hidden_units_0': 230, 'use_batch_norm_0': 0, 'hidden_units_1': 57, 'use_batch_norm_1': 0, 'hidden_units_last': 232, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 782us/step


[I 2022-07-15 19:14:36,589] Trial 73 finished with value: 0.4354991054534912 and parameters: {'n_layers': 3, 'hidden_units_0': 248, 'use_batch_norm_0': 0, 'hidden_units_1': 132, 'use_batch_norm_1': 0, 'hidden_units_last': 242, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 748us/step


[I 2022-07-15 19:15:42,813] Trial 74 finished with value: 0.4377467368199275 and parameters: {'n_layers': 3, 'hidden_units_0': 249, 'use_batch_norm_0': 0, 'hidden_units_1': 133, 'use_batch_norm_1': 0, 'hidden_units_last': 242, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 19:16:46,759] Trial 75 finished with value: 0.44834492536691517 and parameters: {'n_layers': 3, 'hidden_units_0': 248, 'use_batch_norm_0': 0, 'hidden_units_1': 130, 'use_batch_norm_1': 0, 'hidden_units_last': 195, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 806us/step


[I 2022-07-15 19:17:53,651] Trial 76 finished with value: 0.4555663585662842 and parameters: {'n_layers': 3, 'hidden_units_0': 255, 'use_batch_norm_0': 0, 'hidden_units_1': 128, 'use_batch_norm_1': 0, 'hidden_units_last': 215, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 809us/step


[I 2022-07-15 19:19:01,508] Trial 77 finished with value: 0.44526439666748047 and parameters: {'n_layers': 3, 'hidden_units_0': 240, 'use_batch_norm_0': 0, 'hidden_units_1': 104, 'use_batch_norm_1': 0, 'hidden_units_last': 223, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 19:19:27,551] Trial 78 finished with value: 0.6773956922384409 and parameters: {'n_layers': 3, 'hidden_units_0': 227, 'use_batch_norm_0': 0, 'hidden_units_1': 155, 'use_batch_norm_1': 0, 'hidden_units_last': 183, 'optimizer': 'AdamW', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 947us/step


[I 2022-07-15 19:20:51,983] Trial 79 finished with value: 0.509604125389686 and parameters: {'n_layers': 3, 'hidden_units_0': 241, 'use_batch_norm_0': 0, 'hidden_units_1': 141, 'use_batch_norm_1': 1, 'hidden_units_last': 243, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 911us/step


[I 2022-07-15 19:22:02,366] Trial 80 finished with value: 0.4790204701056847 and parameters: {'n_layers': 3, 'hidden_units_0': 248, 'use_batch_norm_0': 0, 'hidden_units_1': 148, 'use_batch_norm_1': 0, 'hidden_units_last': 239, 'optimizer': 'Yogi', 'loss': 'mape'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 848us/step


[I 2022-07-15 19:23:11,674] Trial 81 finished with value: 0.45232235468350923 and parameters: {'n_layers': 3, 'hidden_units_0': 234, 'use_batch_norm_0': 0, 'hidden_units_1': 137, 'use_batch_norm_1': 0, 'hidden_units_last': 228, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 748us/step


[I 2022-07-15 19:24:20,662] Trial 82 finished with value: 0.4817424407372108 and parameters: {'n_layers': 3, 'hidden_units_0': 235, 'use_batch_norm_0': 0, 'hidden_units_1': 238, 'use_batch_norm_1': 0, 'hidden_units_last': 35, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 768us/step


[I 2022-07-15 19:25:28,132] Trial 83 finished with value: 0.42434317148648776 and parameters: {'n_layers': 3, 'hidden_units_0': 214, 'use_batch_norm_0': 0, 'hidden_units_1': 111, 'use_batch_norm_1': 0, 'hidden_units_last': 238, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 871us/step


[I 2022-07-15 19:26:32,356] Trial 84 finished with value: 0.4629363184708815 and parameters: {'n_layers': 3, 'hidden_units_0': 216, 'use_batch_norm_0': 0, 'hidden_units_1': 106, 'use_batch_norm_1': 0, 'hidden_units_last': 212, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 19:27:38,508] Trial 85 finished with value: 0.44831781240609975 and parameters: {'n_layers': 3, 'hidden_units_0': 251, 'use_batch_norm_0': 0, 'hidden_units_1': 117, 'use_batch_norm_1': 0, 'hidden_units_last': 220, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 797us/step


[I 2022-07-15 19:28:45,975] Trial 86 finished with value: 0.4599970127986028 and parameters: {'n_layers': 3, 'hidden_units_0': 224, 'use_batch_norm_0': 0, 'hidden_units_1': 94, 'use_batch_norm_1': 0, 'hidden_units_last': 239, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 19:29:53,646] Trial 87 finished with value: 0.44305618542891284 and parameters: {'n_layers': 3, 'hidden_units_0': 208, 'use_batch_norm_0': 0, 'hidden_units_1': 125, 'use_batch_norm_1': 0, 'hidden_units_last': 235, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 19:30:56,137] Trial 88 finished with value: 0.5107461811945988 and parameters: {'n_layers': 3, 'hidden_units_0': 54, 'use_batch_norm_0': 0, 'hidden_units_1': 112, 'use_batch_norm_1': 0, 'hidden_units_last': 196, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 840us/step


[I 2022-07-15 19:31:15,069] Trial 89 finished with value: 0.6454327084467961 and parameters: {'n_layers': 3, 'hidden_units_0': 194, 'use_batch_norm_0': 0, 'hidden_units_1': 134, 'use_batch_norm_1': 0, 'hidden_units_last': 226, 'optimizer': 'AdamW', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 801us/step


[I 2022-07-15 19:32:23,462] Trial 90 finished with value: 0.4549373685396635 and parameters: {'n_layers': 3, 'hidden_units_0': 169, 'use_batch_norm_0': 0, 'hidden_units_1': 248, 'use_batch_norm_1': 0, 'hidden_units_last': 73, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 782us/step


[I 2022-07-15 19:33:26,263] Trial 91 finished with value: 0.4532903143075796 and parameters: {'n_layers': 3, 'hidden_units_0': 231, 'use_batch_norm_0': 0, 'hidden_units_1': 207, 'use_batch_norm_1': 0, 'hidden_units_last': 244, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 19:34:26,866] Trial 92 finished with value: 0.44812544969412 and parameters: {'n_layers': 3, 'hidden_units_0': 241, 'use_batch_norm_0': 0, 'hidden_units_1': 154, 'use_batch_norm_1': 0, 'hidden_units_last': 251, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 798us/step


[I 2022-07-15 19:35:33,304] Trial 93 finished with value: 0.44699207379267764 and parameters: {'n_layers': 3, 'hidden_units_0': 256, 'use_batch_norm_0': 0, 'hidden_units_1': 172, 'use_batch_norm_1': 0, 'hidden_units_last': 232, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 698us/step


[I 2022-07-15 19:36:39,780] Trial 94 finished with value: 0.45473603908832255 and parameters: {'n_layers': 3, 'hidden_units_0': 245, 'use_batch_norm_0': 0, 'hidden_units_1': 216, 'use_batch_norm_1': 0, 'hidden_units_last': 243, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 948us/step


[I 2022-07-15 19:37:49,279] Trial 95 finished with value: 0.4764426411115206 and parameters: {'n_layers': 3, 'hidden_units_0': 227, 'use_batch_norm_0': 0, 'hidden_units_1': 228, 'use_batch_norm_1': 0, 'hidden_units_last': 252, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 698us/step


[I 2022-07-15 19:38:56,289] Trial 96 finished with value: 0.45418522908137393 and parameters: {'n_layers': 3, 'hidden_units_0': 219, 'use_batch_norm_0': 0, 'hidden_units_1': 237, 'use_batch_norm_1': 0, 'hidden_units_last': 179, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 748us/step


[I 2022-07-15 19:40:03,374] Trial 97 finished with value: 0.4569108985020564 and parameters: {'n_layers': 3, 'hidden_units_0': 237, 'use_batch_norm_0': 0, 'hidden_units_1': 189, 'use_batch_norm_1': 0, 'hidden_units_last': 218, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 748us/step


[I 2022-07-15 19:41:12,424] Trial 98 finished with value: 0.4849989898388202 and parameters: {'n_layers': 3, 'hidden_units_0': 213, 'use_batch_norm_0': 0, 'hidden_units_1': 225, 'use_batch_norm_1': 0, 'hidden_units_last': 256, 'optimizer': 'Yogi', 'loss': 'mape'}. Best is trial 55 with value: 0.4111724328994751.


21/21 [==============================] - 0s 731us/step


[I 2022-07-15 19:41:49,738] Trial 99 finished with value: 0.5603517620380108 and parameters: {'n_layers': 1, 'hidden_units_last': 234, 'optimizer': 'Yogi', 'loss': 'mae'}. Best is trial 55 with value: 0.4111724328994751.


number     value             datetime_start          datetime_complete  \
0        0  0.538595 2022-07-15 18:07:44.978654 2022-07-15 18:08:40.118995   
1        1  5.213333 2022-07-15 18:08:40.118995 2022-07-15 18:09:06.696598   
2        2  1.860014 2022-07-15 18:09:06.697602 2022-07-15 18:09:19.100776   
3        3  0.531550 2022-07-15 18:09:19.101778 2022-07-15 18:10:02.722053   
4        4  0.662501 2022-07-15 18:10:02.723051 2022-07-15 18:10:24.834547   
..     ...       ...                        ...                        ...   
95      95  0.476443 2022-07-15 19:36:39.780421 2022-07-15 19:37:49.279034   
96      96  0.454185 2022-07-15 19:37:49.279034 2022-07-15 19:38:56.288019   
97      97  0.456911 2022-07-15 19:38:56.289019 2022-07-15 19:40:03.373852   
98      98  0.484999 2022-07-15 19:40:03.374854 2022-07-15 19:41:12.424663   
99      99  0.560352 2022-07-15 19:41:12.424663 2022-07-15 19:41:49.737891   

                 duration  params_hidden_units_0  params_hidden_units_1  \
0  0 days 00:00:55.140341                  169.0                    NaN   
1  0 days 00:00:26.577603                    NaN                    NaN   
2  0 days 00:00:12.403174                  209.0                    NaN   
3  0 days 00:00:43.620275                    NaN                    NaN   
4  0 days 00:00:22.111496                   74.0                  145.0   
..                    ...                    ...                    ...   
95 0 days 00:01:09.498613                  227.0                  228.0   
96 0 days 00:01:07.008985                  219.0                  237.0   
97 0 days 00:01:07.084833                  237.0                  189.0   
98 0 days 00:01:09.049809                  213.0                  225.0   
99 0 days 00:00:37.313228                    NaN                    NaN   

    params_hidden_units_last params_loss  params_n_layers params_optimizer  \
0                        208         mae                2             Yogi   
1                         44        mape                1            AdamW   
2                         51        mape                2            AdamW   
3                        156         mae                1             Yogi   
4                        126         mae                3            AdamW   
..                       ...         ...              ...              ...   
95                       252         mae                3             Yogi   
96                       179         mae                3             Yogi   
97                       218         mae                3             Yogi   
98                       256        mape                3             Yogi   
99                       234         mae                1             Yogi   

    params_use_batch_norm_0  params_use_batch_norm_1     state  
0                       0.0                      NaN  COMPLETE  
1                       NaN                      NaN  COMPLETE  
2                       0.0                      NaN  COMPLETE  
3                       NaN                      NaN  COMPLETE  
4                       1.0                      0.0  COMPLETE  
..                      ...                      ...       ...  
95                      0.0                      0.0  COMPLETE  
96                      0.0                      0.0  COMPLETE  
97                      0.0                      0.0  COMPLETE  
98                      0.0                      0.0  COMPLETE  
99                      NaN                      NaN  COMPLETE  

[100 rows x 14 columns]

# Construct Best Model

In [9]:
model = create_model(study.best_trial)
model.summary()

model.fit(
    x_train,
    y_train,
    batch_size=32,
    epochs=100,
    callbacks=[early_stopping],
    validation_data=(x_test, y_test),
    verbose=1
)

Model: "sequential_110"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_404 (Dense)           (None, 240)               3120      
                                                                 
 dense_405 (Dense)           (None, 219)               52779     
                                                                 
 dense_406 (Dense)           (None, 251)               55220     
                                                                 
 dense_407 (Dense)           (None, 1)                 252       
                                                                 
Total params: 111,371
Trainable params: 111,371
Non-trainable params: 0
_________________________________________________________________
Epoch 1/100
183/183 [==============================] - 2s 5ms/step - loss: 0.9413 - mae: 0.9413 - val_loss: 0.6409 - val_mae: 0.6409
Epoch 2/100
183/183 [========================

# Generate Predictions

In [27]:
y_pred = model.predict(x_test)
predictions_df = pd.DataFrame({'Ground Truth': y_test, 'Predictions': y_pred.squeeze()})
predictions_df['Predictions'] = predictions_df['Predictions'].apply(lambda pred: round(pred))
predictions_df.to_csv('predictions.csv', index=False, sep=',')
predictions_df

21/21 [==============================] - 0s 2ms/step


Ground Truth  Predictions
5316           6.0            6
5210           6.0            6
3518           5.0            4
1622           6.0            5
2443           5.0            5
...            ...          ...
6486           6.0            6
2805           6.0            7
5681           5.0            5
5771           7.0            7
5545           4.0            5

[650 rows x 2 columns]